### DISCUSSION TITLE
왜 함수형 인터페이스는 `CheckedException` 예외를 던지는 동작을 허용하지 않을까?

### DISCUSSION THREAD

**Discussion by Irisation23** [Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action#36] [CHAPTER 3 람다 표현식 - Q&A] [answered]
### p107 예외, 람다, 함수형 인터페이스의 관계

파트를 확인해 보면 이런 구문이 있다.
```
함수형 인터페이스는 확인된 예외를 던지는 동작을 허용하지 않는다.
```

**CheckedException**은 `I/OException`, `SQLException`등등의 에러가 있다.
왜? **CheckedException**의 동작을 허용하지 않을까?

이말은 `UncheckedException`의 예외 동작은 허용한다것을 의미하는걸까?

**Comment by BvrPark** [comment, marked as answer]
우선 java에는 2가지 종류의 예외가 있다.
### 1. Checked Exception 
Exception 클래스를 상속 받는 예외들로, 메서드가 발생시킬 수 있는 예외를 선언.
Checked Exception을 던질 수 있는 메서드를 호출할 때는 반드시 예외를 처리해야 한다.(ex : throws, try-catch)

### 2. Unchecked Exception
RuntimeException 클래스를 상속받는 예외들이라 예외를 처리하지 않아도 컴파일러에서 오류를 발생시키지 않는다.
그래서 Unchecked Exception을 던질 수 있는 메서드를 호출할 때는 예외 처리를 강제하지 않는다.

이러한 점에 빗대어서
함수형 인터페이스가 Checked Exception을 던지는 동작을 허용하지 않는 이유는
1. Checked Exception은 예외처리를 강제하기 때문에 람다식이 추구하는 간결성과 표현력을 저하시킨다.
2. 함수형 프로그래밍은 단일 책임 원칙을 중요시 하고 그에 따라 함수형 인터페이스는 단 하나의 추상 메서드를 가져야 된다. 
그래서 예외 처리는 해당 메서드의 책임을 초과하는 부분으로 간주될 수 있다.
4. 함수형 인터페이스는 주로 람다식을 통해 익명 함수를 전달하고 실행하는 용도로 사용.
외부에 예외를 던지는 동작은 람다식의 독립성을 훼손시킬 수 있다.

UncheckedException의 예외 동작은 허용한다것을 의미하는걸까?
-> 이 말이 맞다. UncheckedException은 람다식에서 따로 예외를 처리하지 않아도 되기 때문에 UncheckedException의 예외는 허용하는 것이다.

**Reply by Irisation23** [reply]
답변 내용 잘 읽어 봤습니다.
3가지 관점을 토대로 답변을 해 주셨는데 나름대로 해석을 해 봤습니다.

1. 
```
Function<String, String> SomeFunction =  () throws XXXException -> {
  // some Exception Behavior
}
```

2. 은 궁금한 점이 반드시 Exception을 처리 하도록 만든 인터페이스인 `Callable`을 해석 해봐야 할 것 같습니다.

3. 1에 적혀있는 코드의 내용처럼 과연 람다에게 예외를 외부로 던지도록 하는것이 타당한가에 대한 또 다른 견해라고 볼 수 있다고 생각합니다.
